clash-cores/src/Clash/Cores/Xilinx/DcFifo/BlackBoxes.hs

```diff
@@ -0,0 +1,224 @@
+{-|
+  Copyright   :  (C) 2022 Google Inc
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Blackbox implementation for primitives in "Clash.Cores.Xilinx.DcFifo.Explicit".
+-}
+
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Clash.Cores.Xilinx.DcFifo.BlackBoxes where
+
+import Prelude
+
+import Clash.Core.Literal (Literal(NaturalLiteral))
+import Clash.Core.TermLiteral (termToDataError, TermLiteralSNat(..))
+import Clash.Core.Term (Term(Literal))
+import qualified Clash.Primitives.DSL as DSL
+import Clash.Netlist.BlackBox.Types (BlackBoxFunction, emptyBlackBoxMeta)
+import Clash.Netlist.Types (TemplateFunction(..), BlackBox(BBFunction))
+import Clash.Netlist.Util (orNothing)
+
+import Clash.Netlist.Id (Identifier)
+import qualified Clash.Netlist.Id as Id
+import qualified Clash.Netlist.Types as N
+import qualified Clash.Netlist.BlackBox.Types as N
+
+import Control.Monad (unless)
+import Data.Either (lefts)
+import Data.Maybe (catMaybes)
+import GHC.Stack (HasCallStack)
+
+import Clash.Cores.Xilinx.DcFifo.Explicit (DcConfig(..))
+import Clash.Cores.Xilinx.DcFifo.Instances ()
+import Clash.Cores.Xilinx.Common (toTclBool, renderTcl, defIpConfig, IpConfig (properties))
+
+-- | Blackbox function for 'Clash.Cores.Xilinx.Fifo.dcFifo'. It parses the "DcConfig"
+-- supplied to 'dcFifo' from its Term representation, and passes them to two
+-- template functions:
+--
+--   * 'dcFifoTclTF': renders TCL file calling Xilinx's `create_ip`
+--   * 'dcFifoTF': instantiates IP generated in 'dcFifoTclTF'
+--
+-- Additionally, it generates an unique module name for the Xilinx IP.
+--
+dcFifoBBF :: HasCallStack => BlackBoxFunction
+dcFifoBBF _isD _primName args _resTys = do
+  let
+    [  knownNatN, _knownDomainWrite, _knownDomainRead, _knownNatDepth
+     , _constraint1, _constraint2
+     , either error id . termToDataError -> dcConfig
+     , _wClk, _rClk, _rst, _wData
+     , _rEnable
+
+    -- TODO: Make this blackbox return multiple results, instead of a tuple. See:
```

Comment:
Is this done? The PR is merged!

(We return a product type atm).

Reply:
We can leave it for this PR